### PR TITLE
Pricing form updates

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR makes some significant changes to pricing, removing some old (*IMO outdated*) pricing calculations which are:

- Not cached
- Potentially recursive
- Blocking
- Don't add much useful information

The intent here is to move all pricing functionality to the new (cached) scheme, where pricing is available as a *range* which is calculated and updated by the background worker. These old pricing calculations are now a major sticking point for moving to a more API-driven / reactive user interface.

If there are particular features here that get removed, and users have been relying on them, we will need to work out how to make them fit with the new approach.

- Closes https://github.com/inventree/InvenTree/issues/4042
- Closes https://github.com/inventree/InvenTree/issues/4008
- Closes https://github.com/inventree/InvenTree/issues/4009
- Closes https://github.com/inventree/InvenTree/issues/4312

### Tasks

- [x] remove old sales order line item pricing view (super expensive)
- [x] Remove old part-pricing view (modal, slow)
- [x] Remove `get_supplier_price_range`  method
- [x] Remove `get_bom_price_range` method
- [x] Remove `get_price_range` method
- [x] Remove `get_internal_price_range` method
- [x] Remove `get_price_info` method
- [x] Add cache field for `has_complete_bom_price` - https://github.com/inventree/InvenTree/issues/4009
- [ ] Implement business logic for caching 'complete_bom_price' field
- [ ] Data migration for 'complete_bom_price' field
- [ ] Update BOM tables / etc to show complete_bom_price information
- [ ] Add new sales order line pricing (via API)

### CODE REVIEW

File: InvenTree/part/models.py [hunk 1]
```diff
@@ -1628,31 +1627,6 @@ def clear_bom(self):
         """
         self.bom_items.all().delete()
 
-    def getRequiredParts(self, recursive=False, parts=None):
```
Comment (matmair) [inline InvenTree/part/models.py:1631]: Question: Why is this removed? I use it in multiple plugins, is there something wrong with it?

File: InvenTree/part/models.py [hunk 2]
```diff
@@ -1692,147 +1666,6 @@ def schedule_pricing_update(self):
...
-
-        if min_price == max_price:
-            return min_price
-
-        min_price = normalize(min_price)
-        max_price = normalize(max_price)
-
-        return "{a} - {b}".format(a=min_price, b=max_price)
-
-    def get_supplier_price_range(self, quantity=1):
```
Comment (matmair) [inline InvenTree/part/models.py:1719]: Question: What replaces this? A API call?

File: InvenTree/part/models.py [hunk 3]
```diff
@@ -1871,10 +1704,6 @@ def add_price_break(self, quantity, price):
             price=price
         )
 
-    def get_internal_price(self, quantity, moq=True, multiples=True, currency=None):
```
Comment (matmair) [inline InvenTree/part/models.py:1874]: Comment: I am using this in plugins

File: InvenTree/part/models.py [hunk 4]
```diff
@@ -1885,23 +1714,6 @@ def internal_price_breaks(self):
         """Return the associated price breaks in the correct order."""
         return self.internalpricebreaks.order_by('quantity').all()
 
-    def get_purchase_price(self, quantity):
```
Comment (matmair) [inline InvenTree/part/models.py:1888]: Comment: I am using this in plugins

**Review by matmair — COMMENTED** [review]
A pretty big PR; Removes a lot of internal functions I am currently using - how do you imagine them being replaced?
Also, the part pricing modal and line pricing functions are - in my view - integral parts of the pricing function. Having no way to get pricing ranges fast (even if it is blocking) for newly added BOMs removes the main usage for the prototyping guys.

**Comment by SchrodingersGat** [comment]
@matmair  yep a big one. Still trying to work out how to re-introduce those functions in a more "efficient" manner.

It would be easier if we could ignore "quantity" pricing - this is very difficult to cache. And I would argue that because of all the external factors in calculating "quantity" pricing, that there's not much value in InvenTree doing this for the user? I would much rather present a somewhat "fuzzy" range of pricing (based on min/max data) and let the user infer what they will from that.

Thoughts? Probably we have to have a more general conversation about this before I progress this any further.

**Comment by matmair** [comment]
@SchrodingersGat at that point InvenTree becomes less usable than a spreadsheet for pricing. If someone asks how much a part will cost to build at a quantity and they get back a range the tool is not useful, you can not make decisions on that. What sense do price breaks or any price calculations at that point?

**Comment by SchrodingersGat** [comment]
While I don't use it for that purpose, I see your point. However when I do try to use the price-break calculations, it can take tens of seconds to calculate. I'd like (bare minimum) this to be reduced to a more user friendly time. How can we do this do you think?

**Comment by matmair** [comment]
@SchrodingersGat I think with multiple parts there is not much we can do; maybe optimise the database calls? I have no problem moving it to some dedicated page with warning labels or two calculation strategies but outright removing it kneecaps the software IMO.

**Comment by SchrodingersGat** [comment]
@matmair at the moment I cannot think of a good way of making this less expensive. I'll close this for now, maybe revisit at some point.
